README.md
This content was migrated from http://informatics.mayo.edu/svn/trunk/bmi.lexgrid/commons/CTS/CTS/

changelog.txt
CTS Reference Implementation Changelog
*****************************
Release 1.2.13	01/23/07    *
*****************************
* CTS Value Set nesting depth level change.
* Updated to handle a slight change in the index format (full registered name is now stored, 
  instead of just the value after the urn:oid:.
* Updated the test data to have a proper urn:oid instead of urn:iso.
* Added the lvg jar file to the distribution.
* Updated some SQL libraries - fixed a bug that was causing a prepared statement leak.
* Added supported for Prefixed LexGrid SQL Tables.
* hacked code to make MAPI work on newer HL7 DB.  Still a bug that needs to be cleaned up.
* Upgraded to Lucene 2.0.
* Fixed a bug in mapi browser - lookupCodeSystem that wasn't allowing you to pass in null
  for one of the parameters.
* Fixed some of the logger calls that were losing the stack trace information.
* Replaced the phonetix encoder with one from apache commons for doubleMetaphone searches.
* Updated (almost) all libraries to current.
* Updates to match up with the rest of our internal LexGrid environment better.
* Removed LVG Norm - replaced with a Stemming Algorithm that is easier to deploy.
* Upgraded the RIM database to version 0215d.  Updated JUnit tests as necessary.
* Merged the "old" and "new" JUnit tests into the master combined JUnit test suite.
  All VAPI tests now run against the sample data in the LexGrid.xml file.
  All MAPI tests run against the provided RIM database(s).
* Now ship a RIM database in Hypersonic format.  Default MAPI configuration uses the Hypersonic
  database, rather than Access.
* Bug fixes to MAPI implementation to support Hypersonic, MySQL, and Access.
* Underlying LexBIG related code has been updated to the 2006 model format.  CTS has been updated
  so that it can read version 2006 or 2005 XML or SQL tables.  By default, content is loaded into 
  2006 table formats now.
* Added close() methods to the MAPI implementation for internal use.

*****************************
Release 1.2.12	04/11/06    *
*****************************
* Updated EMF / SQL / Converter / Indexer libraries.
* Added support for Oracle to the SQL implementation.
* Added support for newest SQL table structure - forward and backward compatible with old structure.
* Changed the license to EPL
* refactored code to edu.mayo.informatics.cts
* Updated to 1.9.1+ of Lucene - updated to match with changes made to the indexer for LexBIG.
* Split off all of the interface code and HL7 spec into its own project.  Vastly simplified the ant 
  build script.
* Merged the remainder of the CTSUtility project into CTS.  Former CTS changelog is appended to the 
  end of this changelog.

*****************************
Release 1.2.11	11/18/05    *
*****************************
* Fixed a case senstive issue when running on DB2 (sql and sqllite), trying to map an id to a name.
* Fixed a array out of bounds exception that could happen if supported code systems are missing.
* Updating the LexGrid converter libraries for the new LexGrid SQL Schema.  The SQL table changes 
   do not affect CTS.

*****************************
Release 1.2.10	10/17/05     *
*****************************
* New Lucene index format - now one index for regular and norm searching (instead of two)

* Lucene searches now support a "sounds like" algorithm for matching - double metaphone, to be
  specific.  
  
* Changed the setup/configuration loading methods - now supports loading properties configuration 
  files from the classpath - and has a cleaner API.
  
* Added new simpler methods for constructing the implementations.
  
* Made the system cache consistant - and size configurable in the properties file.

* Changed the index structure on the SQL tables - changed the variable order of the SQL queries to match
  (helps performance on SQL dbs with poor optimizers)


*****************************
Release 1.2.9	9/29/05     *
*****************************
* Removed all uses of "Select count(*) > 0 as foo" since it is not supported by DB2

* Fixed DB2 support.

* Updated the GenericSQLConverter code for DB2 issues, consistency.

* Fixed a bug in the LDAP VAPI implementation that was causing it to miss certain relationships
  on a lookupCompleteCoded concept call.

* Query updates for SQLLite to fix DB2 issues, and give performance improvements.

* Overhauled the JUnit test system - now the JUnit tests populate the databases with the necessary
  data prior to the test run, and it is completely dynamic WRT which databases it will test against, 
  and what tests it will run

* Removed the MSAccess LexGrid/LexGrid Lite Rim database - replaced with a HypersonicSQL database.

* Removed the mysql/posgres Rim database dumps - replaced with a LexGrid XML file which can be loaded
  into your database of choice.
  
* Updated structure of shipped libraries, removed unnecessary libraries, files, etc.


*****************************
Release 1.2.8	8/31/05     *
*****************************
* More bug fixes and implementation updates on the LuceneSearch tools.

* Finished the JUnit tests for the LuceneSearch tools

* Fixed a null handling bug in lookupDesignations and lookupProperties across all implementations
if they didn't provide a concept code, or a code system.

* Refactored a few bits of some other JUnit tests.

* Cleaned up some comments/todo items.

* Cleaned up some of the JUnit inconsistencies.

* Updated to latest postgres jdbc driver

* Fixed bugs in both SQL implementations that were allowing more specific language queries to succeed 
on queries that they shouldn't.

* Modified a few of the SQL queries to make better use of indexes when specifiying languages in some queries.

* Fixed a bug in the oldSqlTests.

* Fixed a case-sensitivity bug that affected postgres when trying to find a code system name.


src/org/hl7/utility/VAPIExpansionContext.java
package org.hl7.utility;

import java.io.StringReader;

import org.jdom.Document;
import org.jdom.Element;
import org.jdom.input.SAXBuilder;
import org.jdom.output.Format;
import org.jdom.output.XMLOutputter;

/**
 * <pre>
 * Title:        VAPIExpansionContext
 * Description:  A simple little xml'ized version of the ExpansionContext.
 * Copyright: (c) 2002, 2003, 2004 Mayo Foundation. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification, are
 * permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this list of
 *    conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice, this
 *    list of conditions and the following disclaimer in the documentation and/or other
 *    materials provided with the distribution.
 *
 * 3. The end-user documentation included with the redistribution, if any, must include
 *    the following acknowledgment:
 *       "This product includes software developed by Mayo Clinic Division of Biomedical
 *        Informatics Research (http://informatics.mayo.edu/)."
 *    Alternately, this acknowledgment may appear in the software itself, if and wherever
 *    such third-party acknowledgments normally appear.
 *
 * 4. The names "Mayo", "Mayo Clinic", "Mayo Foundation", or "LexGrid" must not be used 
 *    to endorse or promote products derived from this software without prior written
 *    permission. For written permission, please contact the author or copyright holder.
 *
 * 5. Products derived from this software may not be called "LexGrid", nor may
 *    "LexGrid" or "Mayo" appear in their name, without prior written permission of the
 *    author or copyright holder.
 *
 * THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESSED OR IMPLIED WARRANTIES, INCLUDING,
 * BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL MAYO CLINIC OR OTHER
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 * </pre>
 * 
 * 
 * <H5>Bugs, to be done, notes, etc.</H5>
 * 
 * @author <A HREF="mailto:armbrust.daniel@mayo.edu">Dan Armbrust</A>
 * @version 1.0 - cvs $Revision: 1.10 $ checked in on $Date: 2005/08/04 20:04:29 $
 */

//TODO - This method of parsing the XML is rather slow - experiment with a better way to parse/deparse the XML 
//blobs.  I am seeing upwards of 300 ms just to parse and deparse one of these...  

public class VAPIExpansionContext
{
    public boolean forward_;
    public boolean directDescendantsOnly_;
    public String  ldapName_;
    public String  codeSystemName_;
    public int     nestingDepth_;
    public String  language_;
    public String  defaultLanguage_;
    public String  conceptCode_;
    public String  sourceCodeSystemId_;
    public String  relationship_;
    public String  defaultTargetCodeSystemId_;
    public int timeout_;
    public int sizeLimit_;
    
    private static String formatVersion = "3";

    public VAPIExpansionContext(boolean forward, boolean directDescendantsOnly, String relationship, String ldapName, String codeSystemName, String sourceCodeSystemId, int nestingDepth, String language, String defaultLanguage, String conceptCode,
             String targetCodeSystem, int timeout, int sizeLimit)
    {
        this.forward_ = forward;
        this.directDescendantsOnly_ = directDescendantsOnly;
        this.ldapName_ = ldapName;
        this.codeSystemName_ = codeSystemName;
        this.nestingDepth_ = nestingDepth;
        this.language_ = language;
        this.conceptCode_ = conceptCode;
        this.relationship_ = relationship;
        this.defaultTargetCodeSystemId_ = targetCodeSystem;
        this.timeout_ = timeout;
        this.sizeLimit_ = sizeLimit;
        this.defaultLanguage_ = defaultLanguage;
        this.sourceCodeSystemId_ = sourceCodeSystemId;
    }

    public static VAPIExpansionContext getExpansionContext(byte[] expansionContext) throws Exception
    {
        return new VAPIExpansionContext(new String(expansionContext));
    }

    public VAPIExpansionContext(String expansionContext) throws Exception
    {
        SAXBuilder builder = new SAXBuilder();
        Document doc = builder.build(new StringReader(expansionContext));
        Element rootElement = doc.getRootElement();

        if (rootElement.getAttribute("version") == null || !rootElement.getAttribute("version").getValue().equals("3"))
        {
            throw new RuntimeException("Cannot parse this expansion context");
        }
        
        forward_ = new Boolean(rootElement.getChildText("forward")).booleanValue();
        directDescendantsOnly_ = new Boolean(rootElement.getChildText("directDescendantsOnly")).booleanValue();
        ldapName_ = rootElement.getChildText("ldapName");
        codeSystemName_ = rootElement.getChildText("codeSystemName"); 
        nestingDepth_ = new Integer(rootElement.getChildText("nestingDepth")).intValue();
        language_ = rootElement.getChildText("language"); 
        defaultLanguage_ = rootElement.getChildText("defaultLanguage"); 
        conceptCode_ = rootElement.getChildText("conceptCode");
        relationship_ = rootElement.getChildText("relationship");
        defaultTargetCodeSystemId_ = rootElement.getChildText("defaultTargetCodeSystemId");
        sourceCodeSystemId_ = rootElement.getChildText("sourceCodeSystemId");
        timeout_ = new Integer(rootElement.getChildText("timeout")).intValue();
        sizeLimit_ = new Integer(rootElement.getChildText("sizeLimit")).intValue(); 
    }

    public String toString()
    {
        Document document = new Document(new Element("context"));
        Element root = document.getRootElement();
        root.setAttribute("version", formatVersion);
        
        Element newElement = new Element("forward");
        newElement.setText(forward_ + "");
        root.addContent(newElement);
        
        newElement = new Element("ldapName");
        newElement.setText(ldapName_);
        root.addContent(newElement);
       
        newElement = new Element("codeSystemName");
        newElement.setText(codeSystemName_);
        root.addContent(newElement);
        
        newElement = new Element("nestingDepth");
        newElement.setText(nestingDepth_ + "");
        root.addContent(newElement);
        
        newElement = new Element("language");
        newElement.setText(language_);
        root.addContent(newElement);
        
        newElement = new Element("conceptCode");
        newElement.setText(conceptCode_);
        root.addContent(newElement);
        
        newElement = new Element("relationship");
        newElement.setText(relationship_);
        root.addContent(newElement);

        newElement = new Element("timeout");
        newElement.setText(timeout_ + "");
        root.addContent(newElement);
        
        newElement = new Element("sizeLimit");
        newElement.setText(sizeLimit_ + "");
        root.addContent(newElement);
        
        newElement = new Element("defaultLanguage");
        newElement.setText(defaultLanguage_);
        root.addContent(newElement);
        
        newElement = new Element("directDescendantsOnly");
        newElement.setText(directDescendantsOnly_ + "");
        root.addContent(newElement);
        
        newElement = new Element("defaultTargetCodeSystemId");
        newElement.setText(defaultTargetCodeSystemId_);
        root.addContent(newElement);
        
        newElement = new Element("sourceCodeSystemId");
        newElement.setText(sourceCodeSystemId_);
        root.addContent(newElement);

        XMLOutputter output = new XMLOutputter(Format.getRawFormat());

        return output.outputString(document);
    }
}